File: desktop/src/runtime/documentRead.ts
export const DOCUMENT_READ_PROTOCOL_VERSION = "document-read/v1";
export const DEFAULT_PREVIEW_DOCUMENT_MAX_BYTES = 20 * 1024 * 1024;
export const DEFAULT_DOCUMENT_CHUNK_BYTES = 256 * 1024;

export type DocumentReadSource = "workspace" | "local-preview" | "tauri";
export type DocumentReadTransport = "whole" | "chunked";
export type DocumentReadErrorCode =
  | "not_found"
  | "too_large"
  | "unsupported_encoding"
  | "revision_conflict"
  | "changed_during_read"
  | "cancelled"
  | "invalid_request"
  | "invalid_chunk"
  | "missing_chunks"
  | "io_error";

export interface DocumentReadRequest {
  readonly protocol_version: string;
  readonly request_id: string;
  readonly document_id: string;
  readonly source: DocumentReadSource;
  readonly path: string;
  readonly expected_revision?: string | null;
  readonly preferred_transport: "auto" | DocumentReadTransport;
  readonly chunk_size_bytes: number;
  readonly max_bytes: number;
}

export interface DocumentReadStartMessage {
  readonly protocol_version: string;
  readonly type: "start";
  readonly request_id: string;
  readonly document_id: string;
  readonly source: DocumentReadSource;
  readonly path: string;
  readonly revision: string;
  readonly encoding: "utf-8";
  readonly transport: DocumentReadTransport;
  readonly total_bytes: number;
  readonly chunk_size_bytes: number;
  readonly chunk_count: number;
}

export interface DocumentReadChunkMessage {
  readonly protocol_version: string;
  readonly type: "chunk";
  readonly request_id: string;
  readonly document_id: string;
  readonly revision: string;
  readonly chunk_index: number;
  readonly offset_bytes: number;
  readonly byte_length: number;
  readonly content: string;
}

export interface DocumentReadCompleteMessage {
  readonly protocol_version: string;
  readonly type: "complete";
  readonly request_id: string;
  readonly document_id: string;
  readonly revision: string;
  readonly total_bytes: number;
  readonly chunk_count: number;
}

export interface DocumentReadErrorMessage {
  readonly protocol_version: string;
  readonly type: "error";
  readonly request_id: string;
  readonly document_id: string;
  readonly revision?: string | null;
  readonly code: DocumentReadErrorCode;
  readonly message: string;
  readonly retryable: boolean;
}

export type DocumentReadMessage =
  | DocumentReadStartMessage
  | DocumentReadChunkMessage
  | DocumentReadCompleteMessage
  | DocumentReadErrorMessage;

export interface DocumentReadResult {
  readonly document_id: string;
  readonly source: DocumentReadSource;
  readonly path: string;
  readonly revision: string;
  readonly encoding: "utf-8";
  readonly total_bytes: number;
  readonly content: string;
}

export class DocumentReadProtocolError extends Error {
  constructor(
    readonly code: DocumentReadErrorCode,
    message: string,
    readonly retryable = false,
  ) {
    super(message);
    this.name = "DocumentReadProtocolError";
  }
}

export class DocumentReadAssembler {
  private startMessage: DocumentReadStartMessage | null = null;
  private readonly chunks = new Map<number, DocumentReadChunkMessage>();
  private result: DocumentReadResult | null = null;
  private cancelled = false;

  constructor(readonly request: DocumentReadRequest) {
    validateRequest(request);
  }

  push(message: DocumentReadMessage): "accepted" | "duplicate" | DocumentReadResult {
    this.assertActive();
    validateEnvelope(this.request, message);
    if (message.type === "error") {
      throw new DocumentReadProtocolError(message.code, message.message, message.retryable);
    }
    if (message.type === "start") {
      return this.acceptStart(message);
    }
    const start = this.requireStart();
    validateMessageRevision(start, message);
    if (message.type === "chunk") {
      return this.acceptChunk(start, message);
    }
    return this.acceptComplete(start, message);
  }

  cancel(message = "Document preview read cancelled"): DocumentReadErrorMessage {
    this.cancelled = true;
    return Object.freeze({
      protocol_version: DOCUMENT_READ_PROTOCOL_VERSION,
      type: "error",
      request_id: this.request.request_id,
      document_id: this.request.document_id,
      revision: this.startMessage?.revision ?? null,
      code: "cancelled",
      message,
      retryable: true,
    });
  }

  get completed(): DocumentReadResult | null {
    return this.result;
  }

  private acceptStart(message: DocumentReadStartMessage): "accepted" | "duplicate" {
    validateStart(this.request, message);
    if (this.startMessage) {
      if (sameStart(this.startMessage, message)) return "duplicate";
      throw new DocumentReadProtocolError("invalid_request", "Conflicting document read start message");
    }
    this.startMessage = message;
    return "accepted";
  }

  private acceptChunk(
    start: DocumentReadStartMessage,
    message: DocumentReadChunkMessage,
  ): "accepted" | "duplicate" {
    validateChunk(start, message);
    const existing = this.chunks.get(message.chunk_index);
    if (existing) {
      if (sameChunk(existing, message)) return "duplicate";
      throw new DocumentReadProtocolError("invalid_chunk", `Conflicting duplicate chunk ${message.chunk_index}`);
    }
    this.chunks.set(message.chunk_index, message);
    return "accepted";
  }

  private acceptComplete(
    start: DocumentReadStartMessage,
    message: DocumentReadCompleteMessage,
  ): DocumentReadResult {
    if (message.total_bytes !== start.total_bytes || message.chunk_count !== start.chunk_count) {
      throw new DocumentReadProtocolError("changed_during_read", "Document size or chunk count changed during read", true);
    }
    const missing = Array.from({ length: start.chunk_count }, (_value, index) => index)
      .filter((index) => !this.chunks.has(index));
    if (missing.length) {
      throw new DocumentReadProtocolError("missing_chunks", `Missing document chunks: ${missing.join(",")}`, true);
    }
    const ordered = Array.from(this.chunks.values()).sort((left, right) => left.chunk_index - right.chunk_index);
    let offset = 0;
    for (const chunk of ordered) {
      if (chunk.offset_bytes !== offset) {
        throw new DocumentReadProtocolError("invalid_chunk", `Chunk ${chunk.chunk_index} has a non-contiguous offset`);
      }
      offset += chunk.byte_length;
    }
    if (offset !== start.total_bytes) {
      throw new DocumentReadProtocolError("missing_chunks", `Assembled ${offset} of ${start.total_bytes} bytes`, true);
    }
    const content = ordered.map((chunk) => chunk.content).join("");
    if (utf8Length(content) !== start.total_bytes) {
      throw new DocumentReadProtocolError("invalid_chunk", "Assembled UTF-8 byte length does not match start message");
    }
    this.result = Object.freeze({
      document_id: start.document_id,
      source: start.source,
      path: start.path,
      revision: start.revision,
      encoding: start.encoding,
      total_bytes: start.total_bytes,
      content,
    });
    return this.result;
  }

  private requireStart(): DocumentReadStartMessage {
    if (!this.startMessage) {
      throw new DocumentReadProtocolError("invalid_request", "Document chunk arrived before start message");
    }
    return this.startMessage;
  }

  private assertActive(): void {
    if (this.cancelled) {
      throw new DocumentReadProtocolError("cancelled", "Document preview read was cancelled", true);
    }
    if (this.result) {
      throw new DocumentReadProtocolError("invalid_request", "Document preview read is already complete");
    }
  }
}

export function createDocumentReadRequest(
  options: Omit<DocumentReadRequest, "protocol_version" | "chunk_size_bytes" | "max_bytes"> & {
    readonly chunk_size_bytes?: number;
    readonly max_bytes?: number;
  },
): DocumentReadRequest {
  const request = Object.freeze({
    ...options,
    protocol_version: DOCUMENT_READ_PROTOCOL_VERSION,
    chunk_size_bytes: options.chunk_size_bytes ?? DEFAULT_DOCUMENT_CHUNK_BYTES,
    max_bytes: options.max_bytes ?? DEFAULT_PREVIEW_DOCUMENT_MAX_BYTES,
  });
  validateRequest(request);
  return request;
}

export function createDocumentReadMessages(options: {
  readonly request: DocumentReadRequest;
  readonly revision: string;
  readonly content: string;
  readonly transport?: DocumentReadTransport;
}): readonly DocumentReadMessage[] {
  const { request, revision, content } = options;
  validateRequest(request);
  if (!revision.trim()) throw new DocumentReadProtocolError("invalid_request", "Document revision is required");
  const bytes = new TextEncoder().encode(content);
  if (bytes.byteLength > request.max_bytes) {
    throw new DocumentReadProtocolError("too_large", `Document exceeds preview limit ${request.max_bytes}`);
  }
  const transport = options.transport ?? selectTransport(request, bytes.byteLength);
  const rawChunks = transport === "whole"
    ? (bytes.byteLength ? [{ offset: 0, bytes }] : [])
    : splitUtf8Bytes(bytes, request.chunk_size_bytes);
  const start: DocumentReadStartMessage = Object.freeze({
    protocol_version: DOCUMENT_READ_PROTOCOL_VERSION,
    type: "start",
    request_id: request.request_id,
    document_id: request.document_id,
    source: request.source,
    path: request.path,
    revision,
    encoding: "utf-8",
    transport,
    total_bytes: bytes.byteLength,
    chunk_size_bytes: transport === "whole" ? Math.max(1, bytes.byteLength) : request.chunk_size_bytes,
    chunk_count: rawChunks.length,
  });
  const decoder = new TextDecoder("utf-8", { fatal: true });
  const chunks: DocumentReadChunkMessage[] = rawChunks.map((chunk, index) => Object.freeze({
    protocol_version: DOCUMENT_READ_PROTOCOL_VERSION,
    type: "chunk",
    request_id: request.request_id,
    document_id: request.document_id,
    revision,
    chunk_index: index,
    offset_bytes: chunk.offset,
    byte_length: chunk.bytes.byteLength,
    content: decoder.decode(chunk.bytes),
  }));
  const complete: DocumentReadCompleteMessage = Object.freeze({
    protocol_version: DOCUMENT_READ_PROTOCOL_VERSION,
    type: "complete",
    request_id: request.request_id,
    document_id: request.document_id,
    revision,
    total_bytes: bytes.byteLength,
    chunk_count: chunks.length,
  });
  return Object.freeze([start, ...chunks, complete]);
}

export function createWholeDocumentReadResult(options: {
  readonly request: DocumentReadRequest;
  readonly revision: string;
  readonly content: string;
  readonly byteLength: number;
}): DocumentReadResult {
  const { request, revision, content, byteLength } = options;
  validateRequest(request);
  if (!revision.trim()) throw new DocumentReadProtocolError("invalid_request", "Document revision is required");
  if (byteLength > request.max_bytes) {
    throw new DocumentReadProtocolError("too_large", `Document exceeds preview limit ${request.max_bytes}`);
  }
  const chunkCount = byteLength === 0 ? 0 : 1;
  const assembler = new DocumentReadAssembler(request);
  assembler.push(Object.freeze({
    protocol_version: DOCUMENT_READ_PROTOCOL_VERSION,
    type: "start",
    request_id: request.request_id,
    document_id: request.document_id,
    source: request.source,
    path: request.path,
    revision,
    encoding: "utf-8",
    transport: "whole",
    total_bytes: byteLength,
    chunk_size_bytes: Math.max(1, byteLength),
    chunk_count: chunkCount,
  }));
  if (chunkCount) {
    assembler.push(Object.freeze({
      protocol_version: DOCUMENT_READ_PROTOCOL_VERSION,
      type: "chunk",
      request_id: request.request_id,
      document_id: request.document_id,
      revision,
      chunk_index: 0,
      offset_bytes: 0,
      byte_length: byteLength,
      content,
    }));
  }
  const result = assembler.push(Object.freeze({
    protocol_version: DOCUMENT_READ_PROTOCOL_VERSION,
    type: "complete",
    request_id: request.request_id,
    document_id: request.document_id,
    revision,
    total_bytes: byteLength,
    chunk_count: chunkCount,
  }));
  if (typeof result !== "object") {
    throw new DocumentReadProtocolError("missing_chunks", "Whole document did not produce a result", true);
  }
  return result;
}

function validateRequest(request: DocumentReadRequest): void {
  if (request.protocol_version !== DOCUMENT_READ_PROTOCOL_VERSION) {
    throw new DocumentReadProtocolError("invalid_request", `Unsupported document read protocol ${request.protocol_version}`);
  }
  if (!request.request_id.trim() || !request.document_id.trim() || !request.path.trim()) {
    throw new DocumentReadProtocolError("invalid_request", "request_id, document_id, and path are required");
  }
  if (!Number.isSafeInteger(request.chunk_size_bytes) || request.chunk_size_bytes < 4) {
    throw new DocumentReadProtocolError("invalid_request", "chunk_size_bytes must be an integer >= 4");
  }
  if (!Number.isSafeInteger(request.max_bytes) || request.max_bytes < 0) {
    throw new DocumentReadProtocolError("invalid_request", "max_bytes must be a non-negative integer");
  }
}

function validateEnvelope(request: DocumentReadRequest, message: DocumentReadMessage): void {
  if (message.protocol_version !== DOCUMENT_READ_PROTOCOL_VERSION) {
    throw new DocumentReadProtocolError("invalid_request", `Unsupported message protocol ${message.protocol_version}`);
  }
  if (message.request_id !== request.request_id || message.document_id !== request.document_id) {
    throw new DocumentReadProtocolError("invalid_request", "Document read message identity mismatch");
  }
}

function validateStart(request: DocumentReadRequest, message: DocumentReadStartMessage): void {
  if ((message.encoding as string).toLowerCase() !== "utf-8") {
    throw new DocumentReadProtocolError("unsupported_encoding", `Unsupported preview encoding ${message.encoding}`);
  }
  if (!message.revision.trim()) {
    throw new DocumentReadProtocolError("invalid_request", "Document revision is required");
  }
  if (request.expected_revision && request.expected_revision !== message.revision) {
    throw new DocumentReadProtocolError("revision_conflict", "Document revision does not match expected revision", true);
  }
  if (!Number.isSafeInteger(message.total_bytes) || message.total_bytes < 0 || message.total_bytes > request.max_bytes) {
    throw new DocumentReadProtocolError("too_large", `Document byte size ${message.total_bytes} exceeds preview contract`);
  }
  if (!Number.isSafeInteger(message.chunk_count) || message.chunk_count < 0) {
    throw new DocumentReadProtocolError("invalid_request", "Invalid document chunk count");
  }
  if (message.transport === "whole" && message.chunk_count !== (message.total_bytes === 0 ? 0 : 1)) {
    throw new DocumentReadProtocolError("invalid_request", "Whole transport must contain zero or one chunk");
  }
}

function validateMessageRevision(
  start: DocumentReadStartMessage,
  message: DocumentReadChunkMessage | DocumentReadCompleteMessage,
): void {
  if (message.revision !== start.revision) {
    throw new DocumentReadProtocolError("changed_during_read", "Document revision changed during read", true);
  }
}

function validateChunk(start: DocumentReadStartMessage, message: DocumentReadChunkMessage): void {
  if (!Number.isSafeInteger(message.chunk_index) || message.chunk_index < 0 || message.chunk_index >= start.chunk_count) {
    throw new DocumentReadProtocolError("invalid_chunk", `Chunk index ${message.chunk_index} is out of range`);
  }
  if (!Number.isSafeInteger(message.offset_bytes) || message.offset_bytes < 0) {
    throw new DocumentReadProtocolError("invalid_chunk", "Chunk offset must be non-negative");
  }
  if (utf8Length(message.content) !== message.byte_length) {
    throw new DocumentReadProtocolError("invalid_chunk", `Chunk ${message.chunk_index} UTF-8 byte length mismatch`);
  }
  if (message.offset_bytes + message.byte_length > start.total_bytes) {
    throw new DocumentReadProtocolError("invalid_chunk", `Chunk ${message.chunk_index} exceeds document byte size`);
  }
}

function splitUtf8Bytes(bytes: Uint8Array, maxBytes: number): Array<{ offset: number; bytes: Uint8Array }> {
  const chunks: Array<{ offset: number; bytes: Uint8Array }> = [];
  let offset = 0;
  while (offset < bytes.byteLength) {
    let end = Math.min(offset + maxBytes, bytes.byteLength);
    while (end < bytes.byteLength && end > offset && isUtf8Continuation(bytes[end])) end -= 1;
    if (end === offset) {
      end = Math.min(offset + 1, bytes.byteLength);
      while (end < bytes.byteLength && isUtf8Continuation(bytes[end])) end += 1;
    }
    chunks.push({ offset, bytes: bytes.slice(offset, end) });
    offset = end;
  }
  return chunks;
}

function selectTransport(request: DocumentReadRequest, totalBytes: number): DocumentReadTransport {
  if (request.preferred_transport !== "auto") return request.preferred_transport;
  return totalBytes <= request.chunk_size_bytes ? "whole" : "chunked";
}

function isUtf8Continuation(value: number): boolean {
  return (value & 0b1100_0000) === 0b1000_0000;
}

function utf8Length(value: string): number {
  let bytes = 0;
  for (let index = 0; index < value.length; index += 1) {
    const code = value.charCodeAt(index);
    if (code <= 0x7f) {
      bytes += 1;
    } else if (code <= 0x7ff) {
      bytes += 2;
    } else if (code >= 0xd800 && code <= 0xdbff) {
      const next = value.charCodeAt(index + 1);
      if (next >= 0xdc00 && next <= 0xdfff) {
        bytes += 4;
        index += 1;
      } else {
        bytes += 3;
      }
    } else {
      bytes += 3;
    }
  }
  return bytes;
}

function sameStart(left: DocumentReadStartMessage, right: DocumentReadStartMessage): boolean {
  return JSON.stringify(left) === JSON.stringify(right);
}

function sameChunk(left: DocumentReadChunkMessage, right: DocumentReadChunkMessage): boolean {
  return left.offset_bytes === right.offset_bytes
    && left.byte_length === right.byte_length
    && left.content === right.content
    && left.revision === right.revision;
}

File: backend/app/api/document_read.py
from __future__ import annotations

import asyncio
import hashlib
import os
import stat
from collections.abc import AsyncIterator, Awaitable, Callable
from dataclasses import dataclass
from enum import StrEnum
from pathlib import Path
from typing import Annotated, BinaryIO, Literal

from pydantic import BaseModel, Field
from starlette.responses import StreamingResponse
from starlette.types import Receive, Scope, Send

DOCUMENT_READ_PROTOCOL_VERSION = "document-read/v1"
DEFAULT_PREVIEW_DOCUMENT_MAX_BYTES = 20 * 1024 * 1024
DEFAULT_DOCUMENT_CHUNK_BYTES = 256 * 1024


class DocumentReadSource(StrEnum):
    WORKSPACE = "workspace"
    LOCAL_PREVIEW = "local-preview"
    TAURI = "tauri"


class DocumentReadTransport(StrEnum):
    WHOLE = "whole"
    CHUNKED = "chunked"


class DocumentReadErrorCode(StrEnum):
    NOT_FOUND = "not_found"
    TOO_LARGE = "too_large"
    UNSUPPORTED_ENCODING = "unsupported_encoding"
    REVISION_CONFLICT = "revision_conflict"
    CHANGED_DURING_READ = "changed_during_read"
    CANCELLED = "cancelled"
    INVALID_REQUEST = "invalid_request"
    INVALID_CHUNK = "invalid_chunk"
    MISSING_CHUNKS = "missing_chunks"
    IO_ERROR = "io_error"


class DocumentReadRequest(BaseModel):
    protocol_version: Literal["document-read/v1"] = DOCUMENT_READ_PROTOCOL_VERSION
    request_id: str = Field(min_length=1)
    document_id: str = Field(min_length=1)
    source: DocumentReadSource
    path: str = Field(min_length=1)
    expected_revision: str | None = None
    preferred_transport: Literal["auto", "whole", "chunked"] = "auto"
    chunk_size_bytes: int = Field(default=DEFAULT_DOCUMENT_CHUNK_BYTES, ge=4)
    max_bytes: int = Field(default=DEFAULT_PREVIEW_DOCUMENT_MAX_BYTES, ge=0)


class DocumentReadStartMessage(BaseModel):
    protocol_version: Literal["document-read/v1"] = DOCUMENT_READ_PROTOCOL_VERSION
    type: Literal["start"] = "start"
    request_id: str
    document_id: str
    source: DocumentReadSource
    path: str
    revision: str
    encoding: Literal["utf-8"] = "utf-8"
    transport: DocumentReadTransport
    total_bytes: int = Field(ge=0)
    chunk_size_bytes: int = Field(ge=1)
    chunk_count: int = Field(ge=0)


class DocumentReadChunkMessage(BaseModel):
    protocol_version: Literal["document-read/v1"] = DOCUMENT_READ_PROTOCOL_VERSION
    type: Literal["chunk"] = "chunk"
    request_id: str
    document_id: str
    revision: str
    chunk_index: int = Field(ge=0)
    offset_bytes: int = Field(ge=0)
    byte_length: int = Field(ge=0)
    content: str


class DocumentReadCompleteMessage(BaseModel):
    protocol_version: Literal["document-read/v1"] = DOCUMENT_READ_PROTOCOL_VERSION
    type: Literal["complete"] = "complete"
    request_id: str
    document_id: str
    revision: str
    total_bytes: int = Field(ge=0)
    chunk_count: int = Field(ge=0)


class DocumentReadErrorMessage(BaseModel):
    protocol_version: Literal["document-read/v1"] = DOCUMENT_READ_PROTOCOL_VERSION
    type: Literal["error"] = "error"
    request_id: str
    document_id: str
    revision: str | None = None
    code: DocumentReadErrorCode
    message: str
    retryable: bool


DocumentReadMessage = Annotated[
    DocumentReadStartMessage
    | DocumentReadChunkMessage
    | DocumentReadCompleteMessage
    | DocumentReadErrorMessage,
    Field(discriminator="type"),
]


@dataclass(frozen=True, slots=True)
class DocumentReadSnapshot:
    path: str
    revision: str
    data: bytes


class DocumentReadSnapshotError(Exception):
    def __init__(
        self,
        code: DocumentReadErrorCode,
        message: str,
        *,
        retryable: bool,
        details: dict[str, object] | None = None,
    ) -> None:
        super().__init__(message)
        self.code = code
        self.message = message
        self.retryable = retryable
        self.details = details or {}


def read_stable_utf8_document_snapshot(
    target: Path,
    *,
    public_path: str,
    max_bytes: int,
    open_file: Callable[[Path], BinaryIO] | None = None,
    read_open_file: Callable[[BinaryIO, int], bytes] | None = None,
) -> DocumentReadSnapshot:
    opener = open_file or (lambda path: path.open("rb"))
    reader = read_open_file or (lambda handle, limit: handle.read(limit))
    try:
        handle = opener(target)
    except FileNotFoundError as exc:
        raise DocumentReadSnapshotError(
            DocumentReadErrorCode.NOT_FOUND,
            "Document does not exist",
            retryable=False,
        ) from exc
    except PermissionError as exc:
        raise DocumentReadSnapshotError(
            DocumentReadErrorCode.IO_ERROR,
            "Document cannot be opened",
            retryable=False,
            details={"errno": exc.errno},
        ) from exc
    except IsADirectoryError as exc:
        raise DocumentReadSnapshotError(
            DocumentReadErrorCode.INVALID_REQUEST,
            "Document path is not a file",
            retryable=False,
        ) from exc
    except OSError as exc:
        raise DocumentReadSnapshotError(
            DocumentReadErrorCode.IO_ERROR,
            "Document cannot be opened",
            retryable=True,
            details={"errno": exc.errno},
        ) from exc

    try:
        with handle:
            before = os.fstat(handle.fileno())
            if not stat.S_ISREG(before.st_mode):
                raise DocumentReadSnapshotError(
                    DocumentReadErrorCode.INVALID_REQUEST,
                    "Document path is not a regular file",
                    retryable=False,
                )
            if before.st_size > max_bytes:
                raise _snapshot_too_large(before.st_size, max_bytes)
            data = reader(handle, max_bytes + 1)
            after = os.fstat(handle.fileno())
    except DocumentReadSnapshotError:
        raise
    except PermissionError as exc:
        raise DocumentReadSnapshotError(
            DocumentReadErrorCode.IO_ERROR,
            "Document cannot be read",
            retryable=False,
            details={"errno": exc.errno},
        ) from exc
    except OSError as exc:
        raise DocumentReadSnapshotError(
            DocumentReadErrorCode.IO_ERROR,
            "Document cannot be read",
            retryable=True,
            details={"errno": exc.errno},
        ) from exc

    if len(data) > max_bytes:
        raise _snapshot_too_large(len(data), max_bytes)
    try:
        target_after = target.stat()
    except FileNotFoundError as exc:
        raise _snapshot_changed() from exc
    except OSError as exc:
        raise DocumentReadSnapshotError(
            DocumentReadErrorCode.IO_ERROR,
            "Document state cannot be verified",
            retryable=True,
            details={"errno": exc.errno},
        ) from exc
    if _stat_identity(before) != _stat_identity(after) or _stat_identity(after) != _stat_identity(
        target_after
    ):
        raise _snapshot_changed()
    try:
        data.decode("utf-8", errors="strict")
    except UnicodeDecodeError as exc:
        raise DocumentReadSnapshotError(
            DocumentReadErrorCode.UNSUPPORTED_ENCODING,
            "Document is not valid UTF-8 text",
            retryable=False,
        ) from exc
    return DocumentReadSnapshot(
        path=public_path,
        revision="sha256:" + hashlib.sha256(data).hexdigest(),
        data=data,
    )


def _snapshot_too_large(actual_bytes: int, max_bytes: int) -> DocumentReadSnapshotError:
    return DocumentReadSnapshotError(
        DocumentReadErrorCode.TOO_LARGE,
        "Document exceeds the preview read limit",
        retryable=False,
        details={"actual_bytes": actual_bytes, "max_bytes": max_bytes},
    )


def _snapshot_changed() -> DocumentReadSnapshotError:
    return DocumentReadSnapshotError(
        DocumentReadErrorCode.CHANGED_DURING_READ,
        "Document changed while a stable snapshot was being read",
        retryable=True,
    )


def _stat_identity(value: os.stat_result) -> tuple[int, int, int, int]:
    return (value.st_dev, value.st_ino, value.st_size, value.st_mtime_ns)


class DocumentReadStreamingResponse(StreamingResponse):
    """Streaming response compatible with the app's BaseHTTPMiddleware wrapper."""

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        del scope, receive
        try:
            await self.stream_response(send)
        except OSError:
            return
        if self.background is not None:
            await self.background()


def create_document_read_response(
    request: DocumentReadRequest,
    snapshot: DocumentReadSnapshot,
) -> DocumentReadStreamingResponse:
    transport = (
        DocumentReadTransport.WHOLE
        if request.preferred_transport == "whole"
        or (
            request.preferred_transport == "auto"
            and len(snapshot.data) <= request.chunk_size_bytes
        )
        else DocumentReadTransport.CHUNKED
    )
    spans = (
        []
        if not snapshot.data
        else [(0, len(snapshot.data))]
        if transport is DocumentReadTransport.WHOLE
        else utf8_chunk_spans(
            snapshot.data,
            request.chunk_size_bytes,
            validate_utf8=False,
        )
    )
    return DocumentReadStreamingResponse(
        stream_document_read_messages(request, snapshot, transport, spans),
        media_type="application/x-ndjson",
        headers={
            "Cache-Control": "no-store",
            "X-Document-Revision": snapshot.revision,
            "X-Document-Bytes": str(len(snapshot.data)),
        },
    )


async def stream_document_read_messages(
    request: DocumentReadRequest,
    snapshot: DocumentReadSnapshot,
    transport: DocumentReadTransport,
    spans: list[tuple[int, int]],
    should_cancel: Callable[[], Awaitable[bool]] | None = None,
) -> AsyncIterator[str]:
    start_message = DocumentReadStartMessage(
        request_id=request.request_id,
        document_id=request.document_id,
        source=request.source,
        path=snapshot.path,
        revision=snapshot.revision,
        transport=transport,
        total_bytes=len(snapshot.data),
        chunk_size_bytes=(
            max(1, len(snapshot.data))
            if transport is DocumentReadTransport.WHOLE
            else request.chunk_size_bytes
        ),
        chunk_count=len(spans),
    )
    if should_cancel is not None and await should_cancel():
        return
    yield start_message.model_dump_json() + "\n"
    await asyncio.sleep(0)

    for index, (start, end) in enumerate(spans):
        if should_cancel is not None and await should_cancel():
            return
        chunk_message = DocumentReadChunkMessage(
            request_id=request.request_id,
            document_id=request.document_id,
            revision=snapshot.revision,
            chunk_index=index,
            offset_bytes=start,
            byte_length=end - start,
            content=snapshot.data[start:end].decode("utf-8", errors="strict"),
        )
        yield chunk_message.model_dump_json() + "\n"
        await asyncio.sleep(0)

    if should_cancel is not None and await should_cancel():
        return
    complete_message = DocumentReadCompleteMessage(
        request_id=request.request_id,
        document_id=request.document_id,
        revision=snapshot.revision,
        total_bytes=len(snapshot.data),
        chunk_count=len(spans),
    )
    yield complete_message.model_dump_json() + "\n"


def split_utf8_chunks(data: bytes, chunk_size_bytes: int) -> list[tuple[int, str]]:
    return [
        (start, data[start:end].decode("utf-8", errors="strict"))
        for start, end in utf8_chunk_spans(data, chunk_size_bytes)
    ]


def utf8_chunk_spans(
    data: bytes,
    chunk_size_bytes: int,
    *,
    validate_utf8: bool = True,
) -> list[tuple[int, int]]:
    if chunk_size_bytes < 4:
        raise ValueError("chunk_size_bytes must be at least 4")
    if validate_utf8:
        data.decode("utf-8", errors="strict")
    spans: list[tuple[int, int]] = []
    offset = 0
    while offset < len(data):
        end = min(offset + chunk_size_bytes, len(data))
        while end < len(data) and end > offset and data[end] & 0xC0 == 0x80:
            end -= 1
        if end == offset:
            end = min(offset + 1, len(data))
            while end < len(data) and data[end] & 0xC0 == 0x80:
                end += 1
        spans.append((offset, end))
        offset = end
    return spans


def create_document_read_messages(
    *,
    request: DocumentReadRequest,
    revision: str,
    data: bytes,
    transport: DocumentReadTransport | None = None,
) -> list[DocumentReadMessage]:
    if len(data) > request.max_bytes:
        raise ValueError(f"document exceeds preview limit {request.max_bytes}")
    data.decode("utf-8", errors="strict")
    selected = transport or (
        DocumentReadTransport.WHOLE
        if request.preferred_transport == "whole"
        or (request.preferred_transport == "auto" and len(data) <= request.chunk_size_bytes)
        else DocumentReadTransport.CHUNKED
    )
    raw_chunks = (
        [] if not data else [(0, data.decode("utf-8"))]
        if selected == DocumentReadTransport.WHOLE
        else split_utf8_chunks(data, request.chunk_size_bytes)
    )
    start = DocumentReadStartMessage(
        request_id=request.request_id,
        document_id=request.document_id,
        source=request.source,
        path=request.path,
        revision=revision,
        transport=selected,
        total_bytes=len(data),
        chunk_size_bytes=max(1, len(data))
        if selected == DocumentReadTransport.WHOLE
        else request.chunk_size_bytes,
        chunk_count=len(raw_chunks),
    )
    chunks = [
        DocumentReadChunkMessage(
            request_id=request.request_id,
            document_id=request.document_id,
            revision=revision,
            chunk_index=index,
            offset_bytes=offset,
            byte_length=len(content.encode("utf-8")),
            content=content,
        )
        for index, (offset, content) in enumerate(raw_chunks)
    ]
    complete = DocumentReadCompleteMessage(
        request_id=request.request_id,
        document_id=request.document_id,
        revision=revision,
        total_bytes=len(data),
        chunk_count=len(chunks),
    )
    return [start, *chunks, complete]
